compiler/src/main/kotlin/ast/ast.kt

```diff
@@ -74,39 +122,88 @@ data class NegativeExpr(val ctx: CellmataParser.NegativeExprContext, val value:
 
 data class InverseExpr(val ctx: CellmataParser.InverseExprContext, val value: Expr) : Expr()
 
-data class ArrayLookupExpr(val ctx: CellmataParser.ArrayLookupExprContext, val ident: String = MAGIC_UNDEFINED_STRING, val index: Expr) : Expr()
+data class ArrayLookupExpr(
+    val ctx: CellmataParser.ArrayLookupExprContext,
+    val ident: String = MAGIC_UNDEFINED_STRING,
+    val index: Expr
+) : Expr()
 
-data class ArrayBodyExpr(val ctx: CellmataParser.ArrayValueExprContext, val values: List<Expr>) : Expr()
+/**
+ * @param ctx Node in the parse tree this node corresponds to.
+ * @param values Elements of the array.
+ * @param declaredType Type listed before the body/values of the array.
+ */
+data class ArrayBodyExpr(
+    val ctx: CellmataParser.ArrayValueExprContext,
+    val values: List<Expr>,
+    val declaredType: Type? = UncheckedType
+) : Expr()
 
 data class ParenExpr(val ctx: CellmataParser.ParenExprContext, val expr: Expr) : Expr()
 
 data class NamedExpr(val ctx: ParseTree, var ident: String = MAGIC_UNDEFINED_STRING) : Expr()
 
 data class ModuloExpr(val ctx: CellmataParser.ModuloExprContext, val left: Expr, val right: Expr) : Expr()
 
-data class FuncExpr(val ctx: CellmataParser.FuncExprContext, val args: List<Expr>, var ident: String = MAGIC_UNDEFINED_STRING) : Expr()
+/**
+ * Presents a function call in abstract syntax tree
+ */
+data class FuncExpr(
+    val ctx: CellmataParser.FuncExprContext,
+    val args: List<Expr>,
+    var ident: String = MAGIC_UNDEFINED_STRING
+) : Expr()
 
-data class StateIndexExpr(val ctx: CellmataParser.StateIndexExprContext) : Expr()
+data class StateIndexExpr(val ctx: CellmataParser.StateIndexExprContext) : Expr(type = IntegerType)
 
 // Literals
-data class IntLiteral(val ctx: CellmataParser.Integer_literalContext, var value: Int = -1) : Expr()
+/**
+ * Represents a concrete integer value in the abstract syntax tree
+ */
+data class IntLiteral(val ctx: CellmataParser.Integer_literalContext, var value: Int = -1) : Expr(type = IntegerType)
```

Comment:
I assume that `var value: Int = -1` defines the default value. If so, why choose -1, instead of 0?

Reply:
I choice `-1` to because it's different from the default value of a integer in kotlin, and to help aid in spotting errors while debugging. `-1` is intended to serve as indicator to the programmer that something _might_ be up.

Reply:
We could use `Int? = null`, which is a great way to indicate, that the value is not yet parsed. What do you think?

Reply:
I dislike the kotlin norm of using nullable as optional, but I guess I have to accept that it is the correct way to do it in kotlin

Reply:
Now that I think about we should fix this not only for `IntLiteral` but also for `BoolLiteral`, `FloatLiteral`, and `WorldDimension`. That fix should probably be a separate pull request since it's unrelated to documentation.

Reply:
Agreed. Please create an issue for this.

Reply:
I have create issue #60

Reply:
Nice. When @NicEastvillage approves, resolve discussion and merge PR.